---
title: "Metatranscriptomics data processing"
author: "Sudarshan A. Shetty"
date: "`r date()`"
output: 
  html_document: 
    toc: yes
    toc_depth: 2
    toc_float: true
    code_folding: hide  
editor_options: 
  chunk_output_type: console
---

# Mucus Diet based minimal microbiome (MDb-MM-16)study metatranscriptomics analysis  

This document will list the steps for importing output from SAMSA2 pipeline.  
the analysis here is based on triplicates.  
Read outputs of DIAMOND analysis into R. The code follows the R scripts from SAMSA2 pipeline.  

## Load libraries 

```{r setup message=FALSE, warning=FALSE, eval=FALSE}
# rm(list = ls())
# import other necessary packages
suppressPackageStartupMessages({
  library(DESeq2)
  library(tidyverse)
  library(EnhancedVolcano)
  library(ggplot2)
  library(reshape2)
})
```


## Read counts  

Input files here were obtained from SAMSA2 pipeline (run on MIB-WUR server, micro4).

### Get file names  
```{r eval=FALSE}


# GET FILE NAMES
control_files <- list.files(
  pattern = "control_*",
  full.names = T, recursive = FALSE
)

control_names <- ""
for (name in control_files) {
  control_names <-
    c(control_names, unlist(strsplit(
      name,
      split = "_", fixed = TRUE
    ))[2])
}

control_names <- control_names[-1]
control_names_trimmed <- ""
for (name in control_names) {
  control_names_trimmed <-
    c(control_names_trimmed, unlist(strsplit(
      name,
      split = ".", fixed = TRUE
    ))[1])
}

control_names_trimmed <- control_names_trimmed[-1]

# Experimental 48h

exp_files <- list.files(
  pattern = "experimental_*",
  full.names = T,
  recursive = FALSE
)
exp_names <- ""
for (name in exp_files) {
  exp_names <-
    c(exp_names, unlist(strsplit(
      name,
      split = "_", fixed = TRUE
    ))[2])
}
exp_names <- exp_names[-1]
exp_names_trimmed <- ""
for (name in exp_names) {
  exp_names_trimmed <-
    c(exp_names_trimmed, unlist(strsplit(
      name,
      split = ".", fixed = TRUE
    ))[1])
}
exp_names_trimmed <- exp_names_trimmed[-1]

# sanity check
if (length(exp_files) == 0 || length(control_files) == 0) {
  print("\nWARNING: No files found.  Is the directory correct?  Are the files named with 'control_' and 'experimental_' as prefixes?")
  stop()
}

# head(control_names_trimmed)
# head(exp_names_trimmed)
```

### Read files  

```{r eval=FALSE}

# loading the control table
y <- 0
for (x in control_files) {
  y <- y + 1
  if (y == 1) {
    control_table <- read.table(file = x, header = F, quote = "", sep = "\t")
    colnames(control_table) <- c(x, "V2")
    control_table <- control_table[, c(2, 1)]
  }
  if (y > 1) {
    temp_table <- read.table(file = x, header = F, quote = "", sep = "\t")
    colnames(temp_table) <- c(x, "V2")
    temp_table <- temp_table[, c(1, 2)]
    control_table <- merge(control_table, temp_table, by = "V2", all = T)
  }
}

head(control_table)

control_table[is.na(control_table)] <- 0
rownames(control_table) <- control_table$V2

control_table_trimmed <- data.frame(control_table[, -1])

# loading the experimental table
z <- 0
for (x in exp_files) {
  z <- z + 1
  if (z == 1) {
    exp_table <- read.table(file = x, header = F, quote = "", sep = "\t")
    colnames(exp_table) <- c(x, "V2")
    exp_table <- exp_table[, c(2, 1)]
  }
  if (z > 1) {
    temp_table <- read.table(file = x, header = F, quote = "", sep = "\t")
    colnames(temp_table) <- c(x, "V2")
    exp_table <- merge(exp_table, temp_table[, c(1, 2)], by = "V2", all = T)
  }
}
exp_table[is.na(exp_table)] <- 0
rownames(exp_table) <- exp_table$V2
exp_table_trimmed <- exp_table[, -1]
```

Format the two tables.   
### Create main table  
```{r eval=FALSE}

# getting the column names simplified
colnames(control_table_trimmed) <- control_names_trimmed
colnames(exp_table_trimmed) <- exp_names_trimmed

# merging the two tables (time-consuming step)
complete_table <- merge(control_table_trimmed, exp_table_trimmed, by = 0, all = TRUE)
complete_table[is.na(complete_table)] <- 0
# reducing stuff down to avoid duplicates
complete_table <- aggregate(. ~ Row.names, data = complete_table, sum)

rownames(complete_table) <- complete_table$Row.names

complete_table <- complete_table[!(complete_table$Row.names == ""), ]

# removing extra Row.names column
complete_table <- complete_table[, -1]
complete_table.a <- complete_table
# colnames(complete_table.a)
# head(complete_table)
```


Store the complete table for later anlaysis.  

## SynComRawRNA   
The `SynComRawRNA` consist of raw counts from RNAseq for 36 samples (51138 locus tags) stored as `data.frame` object and made available in `syncomR` package.  

### Link gene names with locus tags     
The `SynComRawRNA` consist of raw counts from RNAseq for 36 samples (51138 genes) stored as `data.frame` object and made available in `syncomR` package.  

```{r eval=FALSE}
# Link gene names with locus tags

## Read MDb-MM annotation tables
# dir.create("data_files")
# Read the table with locus tags gene names
gene_table <- read.table(
  file = "data_files/mdbmm_fas_2_tab.txt",
  header = F,
  sep = "\t", quote = ""
)
head(gene_table)

colnames(gene_table) <- c("LocusTag", "GeneName")

head(gene_table)
```

**Add gene names**  

```{r eval=FALSE}
head(complete_table)
complete_table

complete_table$BacterialStrain <- rownames(complete_table)

# Split the locustages in Bacterial strains column
complete_table <- separate(complete_table, BacterialStrain, c("Bacteria", "Locus"),
  extra = "merge", fill = "right"
) %>% # Split the species_Locustag in locus column
  separate(Locus, c("Species", "Locus"),
    extra = "merge", fill = "right"
  ) %>% # Now join genus and specie names
  unite(BacterialStrain, Bacteria, Species, Species) %>% # Remove the locus column
  select(-matches("Locus"))
```

Now using information from gene_table and complete_table, we include the gene names  

```{r eval=FALSE}
# Add new column called locus tag
complete_table$LocusTag <- rownames(complete_table)
# check rows
dim(complete_table)

dim(gene_table)

# gene_table will have more as it was obtained from fastat file
# so we subset to keep only those detected in RNA-seq

gene_table.sub <- subset(gene_table, LocusTag %in% complete_table$LocusTag)
dim(gene_table.sub)

# Now we have same number of rows i.e. locustags in gene_table
complete_table.genename <- merge(complete_table, gene_table.sub, by = "LocusTag")

# head(sorted_locus_results3)
dim(complete_table.genename)


complete_table.genename$GeneName[complete_table.genename$GeneName == "hypothetical protein"] <- NA

# DT::datatable(complete_table.genename)
# sorted_locus_results3$GeneName

complete_table.genename <- complete_table.genename %>%
  mutate(GeneName = ifelse(is.na(GeneName), LocusTag, as.character(GeneName)))

colnames(complete_table.genename)

complete_table.genename$GeneName <- gsub("\"", "", complete_table.genename$GeneName)
```

**Save SynComRawRNA**  

Rownames are *locus_tags*, 39 columns including *LocusTag*, *BacterialStrain* and *GeneName*  

```{r eval=FALSE}

# dir.create("data_rdata")
SynComRawRNA <- complete_table.genename
save(SynComRawRNA, file = "data_rdata/SynComRawRNA.rda")
```

# SynComRNAKEGG   
Rownames are *locus_tags*, 40 columns including *LocusTag*, *BacterialStrain*, *GeneName* and *KEGG IDs as KO numbers*
## Read KEEG BRITE classification  

```{r, warning=FALSE, message=FALSE}

require(data.table)
keg.brite <- fread("data_files/kegg_classifications/KEGG_htext_classificationDb.txt", header = F, sep = "\t")

# head(keg.brite)
# DT::datatable(keg.brite.x)
colnames(keg.brite) <- c("Level_1", "Level_2", "Level_3", "Level_4", "KO")

keg.brite2 <- keg.brite %>%
  mutate(Level_4 = str_replace(Level_4, "\\s", "|")) %>% # Edit to seprate last column.
  separate(Level_4, into = c("KOID", "KOGeneName"), sep = "\\|") # Edit to seprate last column.

# head(keg.brite2)
```

## GhostKola output of Db-MM  
Load the KO annotation for DB-MM locus tags obained form online GHOSTKola annotation. The conmcatenated amino acid fasta of all then strains was uploaded to GhostKola for annotation using default parameters. 

```{r eval=FALSE}

ghostko <- fread("data_files/kegg_classifications/GhostKola_MDbMM_ko_list.txt", header = F, sep = "\t")
# head(ghostko)
colnames(ghostko) <- c("LocusTag", "KO")
# head(ghostko)
```

```{r eval=FALSE}
# Read the raw counts table

# complete_table #<- readRDS("raw_counts_tables.rds")

# head(complete_table)

# filter to keep only those that are annotated in Db-MM
ko_locus <- filter(ghostko, LocusTag %in% complete_table.genename$LocusTag)

# ko_locus <- subset(ghostko, ghostko$LocusTag %in% rownames(complete_table))
```

```{r eval=FALSE}
dim(ko_locus)
dim(complete_table)

# merge the Ghost KO
complete_table.genename.ko <- merge(complete_table.genename, ko_locus, by = "LocusTag")
```

## BRITE classification  
SynComRNAKEGG consists of following columns:  
*KO*, *LocusTag*, *36 samples*, *BacterialStrain*, *GeneName*, *Level_1*, *Level_2*, *Level_3*, *KOID*, *KOGeneName*

```{r, eval=FALSE}
# keg.brite2
keg.brite.sub <- subset(keg.brite2, KO %in% complete_table.genename.ko$KO)

# dim(keg.brite2)
# dim(keg.brite.sub)
# dim(ghostko)
# dim(complete_table.ko)

keg.brite.sub.na <- subset(keg.brite.sub, !(is.na(KO)))
dim(keg.brite.sub.na)

SynComRNAKEGG <- merge(complete_table.genename.ko, keg.brite.sub, by = "KO")

head(SynComRNAKEGG)
colnames(SynComRNAKEGG)
save(SynComRNAKEGG, file = "data_rdata/SynComRNAKEGG.rda")
```

# SynComGMM  

## Prep file for input to gmm species-function anlaysis  

```{r eval=FALSE}

SynComRNAKEGG$BacterialStrain <- SynComRNAKEGG$LocusTag

head(SynComRNAKEGG)

# Since KO can have multiple pathways assinged the SynComRNAKEGG has multiple entries for each locus tags. we will filter it to contain unique Locus tags
SynComRNAKEGG.uniq <- SynComRNAKEGG %>% distinct(LocusTag, .keep_all = TRUE)

# head(SynComRNAKEGG.uniq)

# unique(SynComRNAKEGG.uniq$BacterialStrain)
# write.csv(complete_table.brite, "complete_table_brite_species.csv")
```

### Save file

```{r eval=FALSE}
require(edgeR)

colnames(SynComRNAKEGG.uniq)

complete_table_brite_sum <- SynComRNAKEGG.uniq %>%
  group_by(KO, BacterialStrain) %>%
  summarize(sum(F5T24), sum(F5T48), sum(F5T176), sum(F5T240), sum(F5T264), sum(F6T24), sum(F6T48), sum(F6T176), sum(F6T240), sum(F6T264), sum(F8T24), sum(F8T48), sum(F8T176), sum(F8T240), sum(F8T264), sum(F5T28), sum(F5T32), sum(F5T52), sum(F5T56), sum(F5T74A5), sum(F5T152), sum(F5T248), sum(F6T28), sum(F6T32), sum(F6T52), sum(F6T56), sum(F6T74A5), sum(F6T152), sum(F6T248), sum(F8T28), sum(F8T32), sum(F8T52), sum(F8T56), sum(F8T74A5), sum(F8T152), sum(F8T248))

head(complete_table_brite_sum)
complete_table_brite_sum <- as.data.frame(complete_table_brite_sum)
### GOTO 03_gmm_species_function.rmd
```

### Count normalization  

```{r eval=FALSE}

table_brite <- complete_table_brite_sum[, c(3:38)]
rownames(complete_table_brite_sum) <- paste0("kogene_", "", seq(1:nrow(complete_table_brite_sum)))
rownames(table_brite) <- paste0("kogene_", "", seq(1:nrow(table_brite)))

head(table_brite)
head(table_brite)

cpm_table_brite_count <- as.data.frame(cpm(table_brite, log = FALSE))
# head(cpm_table_brite_count)

cpm_table_brite_count$BacterialStrain <- complete_table_brite_sum$BacterialStrain

cpm_table_brite_count$KO <- complete_table_brite_sum$KO

head(cpm_table_brite_count)
```

```{r eval=FALSE}
colnames(cpm_table_brite_count) <- c(
  "F5T24",
  "F5T48",
  "F5T176",
  "F5T240",
  "F5T264",
  "F6T24",
  "F6T48",
  "F6T176",
  "F6T240",
  "F6T264",
  "F8T24",
  "F8T48",
  "F8T176",
  "F8T240",
  "F8T264",
  "F5T28",
  "F5T32",
  "F5T52",
  "F5T56",
  "F5T74A5",
  "F5T152",
  "F5T248",
  "F6T28",
  "F6T32",
  "F6T52",
  "F6T56",
  "F6T74A5",
  "F6T152",
  "F6T248",
  "F8T28",
  "F8T32",
  "F8T52",
  "F8T56",
  "F8T74A5",
  "F8T152",
  "F8T248",
  "BacterialStrain",
  "KO"
) ### CHANGE ADD ALL SAMPLES NAMES

brite_gmm <-
  as.data.frame(cpm_table_brite_count[, c(
    "KO",
    "BacterialStrain", "F5T24", "F5T48", "F5T176", "F5T240", "F5T264", "F6T24", "F6T48", "F6T176", "F6T240", "F6T264", "F8T24", "F8T48", "F8T176", "F8T240", "F8T264", "F5T28", "F5T32", "F5T52", "F5T56", "F5T74A5", "F5T152", "F5T248", "F6T28", "F6T32", "F6T52", "F5T56", "F6T74A5", "F6T152", "F6T248", "F8T28", "F8T32", "F8T52", "F5T56", "F8T74A5", "F8T152",
    "F8T248"
  )])

# head(brite_gmm)
colnames(brite_gmm) <-
  c("KO", "Species", "F5T24", "F5T48", "F5T176", "F5T240", "F5T264", "F6T24", "F6T48", "F6T176", "F6T240", "F6T264", "F8T24", "F8T48", "F8T176", "F8T240", "F8T264", "F5T28", "F5T32", "F5T52", "F5T56", "F5T74A5", "F5T152", "F5T248", "F6T28", "F6T32", "F6T56", "F6T52", "F6T74A5", "F6T152", "F6T248", "F8T28", "F8T32", "F8T52", "F8T56", "F8T74A5", "F8T152", "F8T248")

head(brite_gmm)
```

```{r eval=FALSE}
brite_gmm <- separate(brite_gmm, Species, c("Bacteria", "Locus"),
  extra = "merge", fill = "right"
)

head(brite_gmm)
brite_gmm <- separate(brite_gmm, Locus, c("Species", "Locus"),
  extra = "merge", fill = "right"
)
head(brite_gmm)

brite_gmm <- brite_gmm %>% unite(Species, Bacteria, Species)
head(brite_gmm)
brite_gmm <- brite_gmm[, -3]
# head(brite_gmm)
write.table(brite_gmm,
  file = "data_files/mdbMM_brite_gmm_server_filtered.tsv",
  quote = FALSE,
  sep = "\t", col.names = NA
)
```

Upload this file `mdbMM_brite_gmm_server_filtered.tsv` to micro3 server.

### Server GMM calculations  

Copy the code chunk below and run in micro3 R

```{r, eval=FALSE}

############
# dat <- read.table("03_gmm_anlaysis/data/brite_gmm_server.tsv", header=T, sep="\t")

library(omixerRpm)
dat <- read.table("mdbMM_brite_gmm_server_filtered.tsv", header = T, sep = "\t")
# meta <- read.table("gmm/metadata.tsv", header=T, sep="\t")

db <- loadDB("GMMs.v1.07")
dat.sub <- dat[, -1]

dat.sub <- dat.sub[, c("Species", "KO", "F5T24", "F5T48", "F5T176", "F5T240", "F5T264", "F6T24", "F6T48", "F6T176", "F6T240", "F6T264", "F8T24", "F8T48", "F8T176", "F8T240", "F8T264", "F5T28", "F5T32", "F5T52", "F5T56", "F5T74A5", "F5T152", "F5T248", "F6T28", "F6T32", "F6T56", "F6T52", "F6T74A5", "F6T152", "F6T248", "F8T28", "F8T32", "F8T52", "F8T56", "F8T74A5", "F8T152", "F8T248")]

mods <- rpm(dat.sub, minimum.coverage = 0.5, annotation = 2, module.db = db)
saveRDS(mods, "mdbmm_gmm_out_apr2020.rds")
###########
```

### Read GMMs  

```{r eval=FALSE}
# the GMM cal was done on server due to issues with java windows pc. the mods_v.rds file was copied to gmm
# read file from server
# copy omixerRpm output mdbmm_gmm_out.rds into data_filesd folder
# mods <- readRDS("03_gmm_anlaysis/data/mods_gmm.rds")
mods <- readRDS("data_files/mdbmm_gmm_out_apr2020.rds")

modsDF <- cbind(mods@annotation, slot(mods, "abundance"))

# write.table(modsDF, file='03_gmm_anlaysis/tables/modsDF_gmm_linux.tsv',
#            quote=FALSE, sep='\t', col.names = NA)

# head(modsDF)
```

### Links Species-Functions     

```{r eval=FALSE}
# modsDF
curatedGMM_names <- read.table("data_files/ModulesCustom/CuratedGMM_names.txt", header = F, sep = "\t")
# tail(curatedGMM_names)

fil_list <- modsDF$Module
curatedGMM_names_filt <- filter(curatedGMM_names, curatedGMM_names$V1 %in% fil_list)

dim(curatedGMM_names_filt)
# curatedGMM_names_filt$V1
curatedGMM_names_filt <- curatedGMM_names_filt[,-3]
colnames(curatedGMM_names_filt) <- c("Module", "Names")
# unique(mat$from)
# dim(pval)
module_df <- merge(modsDF, curatedGMM_names_filt, by.x = "Module")
dim(module_df)
head(module_df)

# count.module <- module_df %>% group_by(Names, Taxon) %>%
#  summarise(mean(B1T24), mean(B2T24), mean(B3T24),mean(B1T48), mean(B2T48), mean(B3T48))

module_df_lng <- melt(module_df)
head(module_df_lng)
count.module <- dcast(module_df_lng, Taxon + Names + Module ~ variable, value.var = "value")

# head(count.module)

colnames(count.module) <- c("Taxon", "Module", "ModuleID", "F8T176", "F6T248", "F8T152", "F6T74.5", "F8T52", "F5T240", "F5T264", "F8T24", "F5T52", "F6T24", "F8T48", "F5T248", "F5T56", "F5T32", "F6T240", "F6T48", "F8T28", "F6T28", "F6T264", "F8T264", "F8T240", "F5T74.5", "F5T152", "F5T176", "F8T74.5", "F6T56", "F5T24", "F8T32", "F6T52", "F8T56", "F6T32", "F5T28", "F6T152", "F5T48", "F8T248", "F6T176") ## Change according to number of  samples

count.module.df1 <- reshape2::melt(count.module)
# head(count.module.df1)
count.module.df1 <- count.module.df1 %>% tidyr::separate(variable, c("Rep", "TimePoint"), "T")
# head(count.module.df1)
# hist(count.module.df1$value)
# osidase.df1 <- osidase.df1 %>% mutate(value = replace_na(value, 0))
# osidase.df1[osidase.df1 ==0]<- NA

# DT::datatable(count.module.df1)

unique(count.module.df1$TimePoint)

# DT::datatable(count.module.df1)
count.module.df1$TimePoint <- as.numeric(count.module.df1$TimePoint)
count.module.df1$Bioreactor <- as.factor(count.module.df1$Rep)

count.module.df1$Bioreactor <- gsub("F5", "Bioreactor A", count.module.df1$Bioreactor)
count.module.df1$Bioreactor <- gsub("F6", "Bioreactor B", count.module.df1$Bioreactor)
count.module.df1$Bioreactor <- gsub("F8", "Bioreactor C", count.module.df1$Bioreactor)
```

Save **SyncomGMM**
```{r eval=FALSE}
count.module.df1$Taxon <- gsub("Lachnospiraceae_bacterium", "Flavonifractor_plautii", count.module.df1$Taxon)

SyncomGMM <- count.module.df1
#saveRDS(SyncomGMM, file = "data_rdata/SynComGMM.rds")

save(SyncomGMM, file = "data_rdata/SynComGMM.rda")
```

```{r}
sessionInfo()
```


