---
title: "DADA2 processing of Db-MM amplicon sequencing"
author: "Sudarshan A. Shetty"
date: "`r date()`"
output: 
  html_document: 
    toc: yes
    toc_depth: 2
    toc_float: true
    code_folding: hide  
editor_options: 
  chunk_output_type: console
---


# Minimal microbiome (MDb-MM-16)study 16S rRNA analysis  

This file contains codes/steps invovled in processing of raw demultiplexed reads using `dada2` (v.1.10.1). The files created during this steps are stored in *03_amplicon/output/rds/* subdirectory.     

## Load libraries  

```{r, message=FALSE, warning=FALSE}
#library("devtools"); 
#devtools::install_github("timelyportfolio/sankeytree")

library(DT) 
library(ggplot2); packageVersion("ggplot2")
library(dada2); packageVersion("dada2")
library(dplyr)
library(microbiome)
library(microbiomeutilities)
#library(phangorn)
#library(ggtree)
#library(DECIPHER)
library(phyloseq)
#library(ggrepel)

dir.create("output")
dir.create(output/rds)
```


Copied the chunck from DADA2 tutorial and modified to this specific analysis.  

## File parsing

The fastq files were demultiplex using QIMME galaxy version and stored in `amp_seq/demux_files` folder.  

```{r}

path <- "demux_files" # CHANGE ME to the directory containing your demultiplexed fastq files
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_f.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_r.fastq.gz", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_f"), `[`, 1)

length(fnFs)
length(fnRs)

plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(120,120),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
saveRDS(out, "output/rds/filterAndTrim_out.rds")
out <- readRDS("output/rds/filterAndTrim_out.rds")
out.df <- as.data.frame(out)
out.df$PercentKept <- (out.df$reads.out/out.df$reads.in)*100
write.table(out.df, "filter_trim_stats.txt", sep="\t", row.names= T)
summary(out.df$PercentKept)
```

## Learn errors and make seq table  

```{r}

set.seed(2412)
# Learn forward error rates
errF <- learnErrors(filtFs, nbases=1e9, multithread=TRUE)
# Learn reverse error rates
errR <- learnErrors(filtRs, nbases=1e9, multithread=TRUE)
# Sample inference and merger of paired-end reads
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names
for(sam in sample.names) {
  cat("Processing:", sam, "\n")
    derepF <- derepFastq(filtFs[[sam]])
    ddF <- dada(derepF, err=errF, multithread=TRUE)
    derepR <- derepFastq(filtRs[[sam]])
    ddR <- dada(derepR, err=errR, multithread=TRUE)
    merger <- mergePairs(ddF, derepF, ddR, derepR, justConcatenate = TRUE)
    mergers[[sam]] <- merger
}
rm(derepF); rm(derepR)

saveRDS(mergers, "output/rds/mergers.rds")
head(mergers)
#mergers <- readRDS("output/rds/mergers.rds")

# Construct sequence table and remove chimeras
seqtab <- makeSequenceTable(mergers)
# write to disk
saveRDS(seqtab, "output/rds/seqtab.rds") # CHANGE ME to where you want sequence table saved

```


## Remove Bimera and Assign taxonomy

```{r}

seqtab <- readRDS("output/rds/seqtab.rds")

set.seed(5546536)
seqtab <- removeBimeraDenovo(seqtab, 
                             method="consensus", 
                             multithread=FALSE)

saveRDS(seqtab, "output/rds/seqtab_binmera.rds")

seqtab <- readRDS("output/rds/seqtab_binmera.rds")
# Assign taxonomy

mdbmmdb <- "database/mdbmm_16s_db.fas"
tax <- assignTaxonomy(seqtab, mdbmmdb, 
                      minBoot=80, 
                      verbose = T,
                      outputBootstraps = F,
                      tryRC=TRUE)
#head(tax)

saveRDS(tax, "output/rds/tax_final.rds") # CHANGE ME ...

#hitdb <- "database/hitdb_v1.00.fa.gz"



```


## Convert to phyloseq  

```{r}

seqtab <- readRDS("output/rds/seqtab_binmera.rds")

tax <- readRDS("output/rds/tax_final.rds") # CH


#seqtab_nochim <- readRDS("output/rds/seqtab_nochim.rds")

#library(phyloseq)
# read metadata table
metatable <- read.table("database/Mucus_minimal_amplicon_mapping_R.txt", 
                      stringsAsFactors = F, row.names = 1, sep = "\t",
                      header = T)
head(metatable)

samdat <- sample_data(metatable)

# make phyloseq object
ps <- phyloseq(otu_table(t(seqtab), taxa_are_rows=T), sample_data(samdat),
               tax_table(tax))

ps


#write_phyloseq(ps, "METADATA")

sample_names(ps) <- sample_data(ps)$FAIR_Labels

# Check for species detected
table(tax_table(ps)[, "Species"], exclude = NA)
saveRDS(ps, "output/rds/ps_fr.rds") 

```

```{r}
sessionInfo()
```



