---
title: "Functional analysis of a synthetic gut microbiome reveals niche segregation as a driving force for assembly, stability, co-existence and activity"
subtitle: "Assembly and co-existence of a butyrate-propionate producing MDb-MM"
author: "Sudarshan A. Shetty"
date: "`r date()`"
output: 
  html_document: 
    toc: yes
    toc_depth: 3
    toc_float: true
    code_folding: show 
    fig_caption: yes 
---

# Investigate metabolite and amplicon data.     

```{r eval=FALSE}

dir.create("ms")
dir.create("ms/figs")
dir.create("ms/tabs")
dir.create("ms/rds")
```

## Load pkgs  
```{r pkgs, message=FALSE, warning=FALSE}
### Setup anlaysis 
suppressPackageStartupMessages({
  library(syncomR)
  library(dplyr)
  library(tidyr)
  library(seqtime)
  library(ggpubr)
  library(scales)
  library(codyn)
  library(knitr)
  library(reshape2)
  library(ggrepel)
  library(patchwork)
  library(qgraph)
  library(DESeq2)
  library(stringr)
  library(ggridges)
})

#devtools::install_github("hallucigenia-sparsa/seqtime")

### Assign colors to key variables  
#fasting_cols <- c("#C71585", "#fb8072", "#80b1d3")
fasting_cols <- c(`DoS pulse`="#C71585",
                  Overnight="#fb8072", 
                  `elongated fasting`="#80b1d3")

fer_cols <- c(`Bioreactor A`= "#b2182b", 
              `Bioreactor B`="#2166ac", 
              `Bioreactor C` = "#35978f")
colorsSCFA <- c(Acetate = "#CBD588", Propionate = "#5F7FC7", 
                Butyrate = "orange", Isobutyrate= "#DA5724", 
                Formate ="#508578", Lactate = "#CD9BCD", 
                Glucose = "#AD6F3B", Fructose = "#673770",
                Succinate = "#D14285")
```


# Results  
The 16S rRNA gene amplicon sequencing data was analyzed using the DADA2 R package.  
A custom database consisting of 16S rRNA gene sequences fetched from the genomes of the 16 bacterial strains used in this study using `barnap` (available at https://github.com/microsud/MDb-MM). Taxonomic assignment was done using the RDP classifier 88. The unique amplicon sequence variants (ASVs) were merged at species level using the `tax_glom()` function in `phyloseq` (v1.32).   

## Experimental design  
**Assembling a minimal microbiome, nutrient periodicity and perturbations**   
During the course of 460h of operation, a defined set of changes were enforced to investigate assembly, compositional and niche dynamics, community stability, etc.  

### Plot central sampling stratergy   

Use info from one bioreactor.  

```{r sampling, fig.height=8, fig.width=16, fig.cap='Bioreactor operation and longitudinal sampling of the minimal microbiome', eval=FALSE, echo=FALSE}

metadf <- subset(meta(SyncomFiltData), StudyIdentifier == "Bioreactor A")

p_batch_cont <- ggplot(metadf) +
  geom_tile(aes(x = as.factor(Time_hr_num), 
                y = StudyIdentifier,
                fill = Type),
            height = 1,color ="white") + 
  theme_minimal(base_size = 6) +
  scale_fill_manual(values = c("#C71585", "#008080")) + 
  xlab("Time (hr)") + rremove("xy.text") + 
  rremove("xy.title") + 
  theme_void()

metadf <- metadf %>% 
  mutate(Condition= ifelse(Fasting=="YES", "Overnight", 
                           ifelse(Fasting=="NO", "DoS pulse","NA"))) %>% 
  mutate(Condition = ifelse(Time_hr_num <= 24, "Batch",Condition))

p_fasting <- ggplot(metadf) +
  geom_tile(aes(x = as.factor(Time_hr_num), 
                y = StudyIdentifier,
                fill = Condition), height = 1,color ="white") +
  theme_minimal(base_size = 6) + 
  rremove("xy.title") + 
  xlab("Time (hr)") + 
  scale_fill_manual(values = c("#C71585", "#fb8072", "#80b1d3")) +
  rremove("xy.text") +theme_void()#+

metadf <- metadf %>% 
  mutate(events= ifelse(Pertubation %in% c("batch_carbs","spike_carbs","fasting"),"Normal", Pertubation))

perturbation_p <- ggplot(metadf) +
  geom_tile(aes(x = as.factor(Time_hr_num), y = StudyIdentifier,
                fill = Pertubation), height = 1,color ="white") + theme_minimal(base_size = 6)+ 
  xlab("Time (hr)") + scale_fill_brewer(palette = "Paired") + rremove("y.title") + rremove("y.text") + theme_void()+ theme(axis.text.x = element_text(size = 6))

########################################## working code end ###################

p_batch_cont / p_fasting / perturbation_p + plot_layout(guides = 'collect') & theme(legend.position = 'bottom')

ggsave("ms/figs/fig_1_sampling_plots.pdf", height = 6, width = 12)
```


## Dynamic assembly and co-existence of strains in MDb-MM  

Checking how many samples for fasting and no fasting for each bioreactor.  
```{r fascond-bio}
table(meta(SyncomFiltData)$Fasting_condition, meta(SyncomFiltData)$StudyIdentifier)
```

### Taxa detections  

Plot presence/absence of taxa at each of the sampling timepoints in the three bioreactors.  

```{r}
ps1.sub.df <- psmelt(SyncomFiltData) %>% 
  arrange(Time_hr_num)

ps1.sub.df.bin <- ps1.sub.df %>% 
  mutate(Detection = ifelse(Abundance >= 1, "Detected", "Not Detected"))

plot_binary <- ps1.sub.df.bin %>% 
  mutate_at(.vars = "OTU", 
            .funs = gsub,
            pattern = "_",
            replacement = " ") %>% 
  ggplot(aes(reorder(Time_hr, Time_hr_num),OTU)) + 
  geom_tile(aes(fill=Detection), color="white") +
  facet_wrap(~StudyIdentifier) + theme_syncom() + 
  scale_fill_manual(values= c("#89b0ae", "#001233")) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(face="italic")) + 
  #theme(panel.grid.major.y = element_blank()) +
  theme(panel.background = element_rect(fill = "#000000")) +
  xlab("Time (h)") + ylab("")
plot_binary 
ggsave("ms/figs/fig_1_presence_absence.pdf", height = 6, width = 12)
```

### Composition of MDb-MM   

```{r fig.height=4, fig.width=9}

data("SyncomFiltData")
taxa_names(SyncomFiltData) <- gsub("_", " ", taxa_names(SyncomFiltData))
syncom.rel <- microbiome::transform(SyncomFiltData, "compositional")
p.comp <- syncomR::plot_syncom_composition(syncom.rel, 
                                           type = "area",
                                           sp.fill.pal = syncom_colors2("Species"), 
                                           facet.var = "StudyIdentifier") +
  theme(legend.key.size = unit(0.5, 'cm'),
        legend.text = element_text(size = rel(0.1)))
p.comp
ggsave("ms/figs/01_fig_MDbMM_composition.pdf", h=4, w=9)
```


### Individual taxa  
Check abundances of individual taxa from time 0h to 460h.     

```{r}
ind_tax_plot <- ps1.sub.df %>% 
  mutate(copies = log10(Abundance +1)) %>% 
  mutate_at(.vars = "OTU", 
            .funs = gsub,
            pattern = "_",
            replacement = " ") %>% 
  ggplot(aes(Time_hr_num, copies)) +
  geom_line(aes(color=StudyIdentifier), alpha=0.5, size=1) +
  facet_wrap(~OTU, scales="free", ncol = 3) + 
  theme_syncom() + 
  scale_color_manual(values = fer_cols) + 
  theme(strip.text = element_text(face = "italic",size=10),
        axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=10),
        legend.text = element_text(size = 10),
        legend.key.size = unit(0.5, 'cm'),
        legend.title = element_text(size = 10)) + 
  geom_vline(xintercept = 152, lty=2, color="#991720") + 
  geom_vline(xintercept = 248, lty=2, color= "#0963bd") +
  ylab("16S rRNA gene copies (log10 +1)") + ylim(0,11) + xlab("Time (hr)")
ggsave("ms/figs/fig_1_taxa_abundances.pdf", height = 6, width = 12)
```

### Major SCFAs  
```{r hplc, fig.align='center', fig.height=4, fig.width=8, fig.cap='SCFA production in MDb-MM'}

#scale_y_continuous(expand = expansion(mult = c(0, 0.1)))

hplc.dat <- system.file("extdata", "hplc_first_check.csv", package = "syncomR")

hplc.dat <- read.csv(hplc.dat, 
                     row.names = 1, 
                     header = T, 
                     stringsAsFactors = F)

hplc.dat <- as.matrix(t(hplc.dat[,c("Acetate", "Butyrate", "Propionate")]))

ps1.hplc <- SyncomFiltData
tax_table(ps1.hplc) <- NULL
otu_table(ps1.hplc) <- otu_table(hplc.dat, taxa_are_rows = T)
hplc.df <- psmelt(ps1.hplc)
colnames(hplc.df)[2] <- "Metabolites"
colnames(hplc.df)[3] <- "Concentration"
hplc.df$Time <- as.numeric(hplc.df$Time_hr_num)

hplc.plot <- ggplot(hplc.df, 
                    aes(Time, Concentration, color=Metabolites)) +
  geom_smooth() +
  geom_point(alpha=0.5) + 
  facet_wrap(~StudyIdentifier, scales = "free", 
             nrow = 3, ncol = 1) + 
  ylab("Concentration (mM)") + 
  xlab("Time (hr)") + 
  scale_color_manual("Metabolites",values = colorsSCFA) + 
  theme_syncom() + 
  geom_vline(xintercept = 152, lty=2, color="#991720") + 
  geom_vline(xintercept = 248, lty=2, color= "#0963bd") +
  theme(axis.ticks.y = element_blank(), 
        axis.text.x = element_text(size=10),
        strip.text = element_text(size=10),
        legend.text = element_text(size = 10),
        legend.key.size = unit(0.5, 'cm'),
        legend.title = element_text(size = 10),
        legend.background=element_rect(fill="white")) +
  guides(color = guide_legend(ncol = 1)) 
#hplc.plot
```

#### Combine strain abundances and SCFA plot 
```{r}
ind_tax_plot + hplc.plot +
  plot_layout(guides = "collect", widths = c(2,1)) + 
  plot_annotation(tag_levels = "A")
ggsave("ms/figs/01_ind_tax_hplc.pdf", height = 6, width = 14)
```


### Biomass differences   

#### O.D  
O.D measurements provided the earliest view of growth response of the MDb-MM to fasting and feeding regimes. The samples were taken before adding carbohydrates for each sampling time.   
```{r biomas-dif, fig.height=3, fig.width=6, warning=FALSE}

od <- meta(SyncomFiltData) %>% 
  select(FAIR_Labels, Fasting, Acetate_Feed,BlautiaHydro,OpticalDensity, Time_hr_num, StudyIdentifier,Fasting_condition, Pertubation) 

od$Time_hr_num <- as.numeric(od$Time_hr_num)
od$OpticalDensity <- as.numeric(od$OpticalDensity)
#od <- subset(od, Time_hr_num < 152)
od <- od %>% mutate(Condition= ifelse(Fasting=="YES", "Overnight", ifelse(Fasting=="NO", "DoS pulse", "NA"))) 

od$Condition <- factor(od$Condition, levels=c( "DoS pulse","Overnight"))

biom.plot <- ggviolin(od, "Condition", "OpticalDensity",
                      color= "Condition",
                      add = "jitter",
                      add.param = list(size= 2, alpha=0.5),
                      palette = fasting_cols
) + 
  stat_compare_means(label = "p.signif",label.y = 11) + 
  theme_syncom() + 
  ylim(1,12) + 
  ylab(expression(O.D[600~nm])) + 
  xlab(NULL) + 
  theme(legend.position = "none") + 
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))
biom.plot
```

#### qPCR  

Check if there is difference in 16S rRNA gene copies between DoS and Overnight samples.  
```{r}

qpcr <- meta(SyncomFiltData) %>% 
  select(FAIR_Labels, Fasting, Acetate_Feed,BlautiaHydro,OpticalDensity, Time_hr_num, StudyIdentifier,Fasting_condition, Pertubation,total_copies) 

qpcr$Time_hr_num <- as.numeric(qpcr$Time_hr_num)
qpcr$total_copies <- as.numeric(qpcr$total_copies)
qpcr$total_copieslog <- log10(qpcr$total_copies)
#od <- subset(od, Time_hr_num < 152)
qpcr <- qpcr %>% mutate(Condition= ifelse(Fasting=="YES", "Overnight", ifelse(Fasting=="NO", "DoS pulse", "NA"))) 

qpcr$Condition <- factor(qpcr$Condition, levels=c( "DoS pulse","Overnight"))

qpcr.plot <- ggviolin(qpcr, "Condition", "total_copieslog",
                      color= "Condition",
                      add = "jitter",
                      add.param = list(size= 2, alpha=0.5),
                      palette = fasting_cols
) + 
  stat_compare_means(label = "p.signif") + 
  theme_syncom() + ylab("Total 16S rRNA copies \n(log10)") + xlab(NULL) + 
  theme(legend.position = "none") + 
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))
qpcr.plot

```


The observation of O.D. was not captured with total 16S rRNA gene qPCR.  

#### Inequality  
Check if there is inequality in abundances, like one strain over-dominating the system?  

```{r}

p.temp.div <- temporal_diversity(SyncomFiltData, 
                                 time.col = 'Time_hr_num', 
                                 div.measure = "gini")

ev.dat <- p.temp.div$data %>% mutate(Condition= ifelse(Fasting=="YES", "Overnight", ifelse(Fasting=="NO", "DoS pulse", "NA"))) 

ev.dat$Condition <- factor(ev.dat$Condition, levels=c( "DoS pulse","Overnight"))
ineq_comp_plot <- ggviolin(ev.dat, "Condition", "Inequality",
                           color= "Condition",
                           add = "jitter",
                           add.param = list(size= 2, alpha=0.5),
                           palette = fasting_cols
) + 
  stat_compare_means(label = "p.signif",label.y = 0.9) + 
  theme_syncom() + ylab("Inequality") + xlab(NULL) + 
  theme(legend.position = "none") + 
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))

#ineq_comp_plot  
```


There seems to be not major difference in evenness in the community.  

#### Coverage    
Check how many strains contribute to 90% of the total abundance.  

```{r coverage, fig.height = 4, fig.width = 4, fig.align = "center", fig.cap="Coverage analysis"}
cov.plot.sub <- taxa_coverage(SyncomFiltData, coverage = 0.90, 
                              time.column = "Time_hr_num", 
                              shape.variable = "Acetate_Feed", 
                              color.variable = "Fasting", 
                              color.pal = fasting_cols) 


cov.dat <- cov.plot.sub$data %>% mutate(Condition= ifelse(Fasting=="YES", "Fasting", ifelse(Fasting=="NO", "Feeding", "NA")))
cov.dat <- cov.dat %>% mutate(Condition= ifelse(Fasting=="YES", "Overnight", ifelse(Fasting=="NO", "DoS pulse", "NA"))) 

cov.dat$Condition <- factor(cov.dat$Condition, levels=c( "DoS pulse","Overnight"))

cov.dat.plot <- ggviolin(cov.dat, "Condition", "diversity",
                         color= "Condition",
                         add = "jitter",
                         add.param = list(size= 1, alpha=0.5),
                         palette = fasting_cols
)+ 
  stat_compare_means(label = "p.signif",label.y = 12) + 
  theme_syncom() + ylab("Coverage") + xlab(NULL)+ 
  theme(legend.position = "none")+ 
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))

```

#### Combine OD, qPCR, inequlaity and coverage.  

```{r}
(biom.plot + qpcr.plot) / (ineq_comp_plot + cov.dat.plot) +  
  plot_layout(guides = "collect") + 
  plot_annotation(tag_levels = "A")

ggsave("ms/figs/01_supp_od_qpcr_gini_cov.pdf", height = 4, width = 8)
ggsave("ms/figs/01_supp_od_qpcr_gini_cov.png", height = 4, width = 8, dpi=600)
```


Check for differences in SCFA in the first 148h.  
Before addition of _B. hydrogenotrophica_
```{r}

hplc.df.148 <- subset(hplc.df, Time_hr_num < 152)

hplc.df.148 <- hplc.df.148 %>% mutate(Condition= ifelse(Fasting=="YES", "Overnight", ifelse(Fasting=="NO", "DoS pulse", "NA"))) 
hplc.df.148 <- hplc.df.148 %>% mutate(Acetate_Conc= ifelse(Acetate_Feed=="YES", "With Acetate", ifelse(Acetate_Feed=="NO", "Without Acetate", "NA"))) 

hplc.df.148$Condition <- factor(hplc.df.148$Condition, levels=c( "DoS pulse","Overnight"))


hplc.148.plot <- ggviolin(hplc.df.148, "Condition", "Concentration",
                          color= "Condition",
                          add = "jitter",
                          add.param = list(size= 2, alpha=0.5),
                          palette = fasting_cols,
                          facet.by = "Metabolites",
                          xlab=""
) + 
  stat_compare_means(label = "p.signif",
                     size=4
                     #label.y = 0.85
  ) + ggtitle("First 148 h")+
  theme_syncom() + ylab("Concentration (mM)")+ 
  theme(legend.position = "none")+ 
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))


```

Check for differences in SCFA after addition of _B. hydrogenotrophica_ and presence/absence of exogenous acetate.

```{r}
hplc.df.152 <- subset(hplc.df, Time_hr_num >= 152)
hplc.df.152 <- hplc.df.152 %>% mutate(Condition= ifelse(Fasting=="YES", "Overnight", ifelse(Fasting=="NO", "DoS pulse", "NA"))) 
hplc.df.152 <- hplc.df.152 %>% mutate(Acetate_Conc= ifelse(Acetate_Feed=="YES", "With Acetate", ifelse(Acetate_Feed=="NO", "Without Acetate", "NA"))) 

hplc.df.152$Condition <- factor(hplc.df.152$Condition, levels=c( "DoS pulse","Overnight"))

hplc.152.plot <- ggviolin(hplc.df.152, "Condition", "Concentration",
                          facet.by = c("Acetate_Conc","Metabolites"), 
                          color= "Condition",
                          palette = fasting_cols,
                          add = "jitter",
                          xlab="",
                          add.param = list(size= 3, alpha=0.5)) + 
  stat_compare_means(label = "p.signif",
                     size=4
                     #label.y = 0.85
  ) + ggtitle("152h and later")+
  theme_syncom() + ylab("Concentration (mM)")+ 
  theme(legend.position = "none")+ 
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))
```

```{r fig.height=10, fig.width=10}
hplc.148.plot / hplc.152.plot + plot_layout(guides = "collect", heights = c(1,2)) + 
  plot_annotation(tag_levels = "A")
ggsave("ms/figs/02a_supp_scfa_subset.pdf", height = 6, width = 8)
ggsave("ms/figs/02a_supp_scfa_subset.png", height = 6, width = 8, dpi=600)
```



```{r}
sessionInfo()
```


